### PR TITLE
Sinowealth based Daly BMS support

### PR DESCRIPTION
**Pull request by Carstijn**
This implements support for Sinowealth chipset based BMS, for now limited to SH367303 / 367305 / 367306 / 39F003 / 39F004 / BMS_10 chips with up to 10 cells. I have tested this with my Daly SmartBMS 4S 250A BMS, this is based on the SH39F003 chipset. My battery bank is a lishen 4s3p, 12 cell 840 Ah 12V bank installed on my boat. Using a solar MPPT and Multiplus 12/3000 for charging and Venus GX device + BMV712 for controlling the setup.

**What works:**
- Reading SOC, Voltage and Current
- Reading Cell Voltages and passing Min and Max values
- Reading capacity and capacity remaining
- Reading BMS status, charge/discharge fets (not all cases tested)
- Reading Alarm flags (not all cases tested)
- Reading temperature

**What kinda works**
- Reading cycles
Reading cycles is implemented, but does not match bluetooth app value

**What is not implemented**
The BMS has more data avaiable in the app, I am not able to read this data atm. It seems there is some way to read  "extended parameter" pages, no clue on how to do this. Because of this some values are hard coded, like cell voltage thresholds and discharge and charge current limits.

**On this pull request**
- Python is not my first language, so codestyle and code improvemt suggestions are welcome.
- It is only tested on my setup, in limited fasion, let me know what works and does not work in your case
- My vacation ends tomorrow, so I have less/no access to the setup on a daily basis and have to see when I am back at my boat to test new changes.
- I am not sure on what needs to be done to complete this pullrequest, so @Louisvdw please fill me in on what is missing.

### CODE REVIEW

File: etc/dbus-serialbattery/utils.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def read_serial_data(command, port, baud, length_pos, length_check, length_fixed
             count = 0
             data = bytearray(res)
             while len(data) <= length + length_check:
-                res = ser.read(length + length_check)
+                res = ser.read((length + length_check) - len(data))
```
Comment (Louisvdw) [inline etc/dbus-serialbattery/utils.py:61]: Can you give me more information on why this change was needed as it could influent the existing BMS that use this function. If you give it more bytes to read it, read will still only read what is available in the buffer. So more is good. 
Did you get any problems in the Sinowealth based Daly using the read without the  -len(data) ?

File: etc/dbus-serialbattery/sinowealth.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        
+        self.cell_max_no = max(self.cell_voltages, key=int) - 1
+        self.cell_max_voltage = max(self.cell_voltages.values())
+        self.cell_min_no = min(self.cell_voltages, key=int) - 1
+        self.cell_min_voltage = min(self.cell_voltages.values())
+        
+        logger.info(">>> INFO: Max cell voltage: %u:%f", self.cell_max_no, self.cell_max_voltage)
+        logger.info(">>> INFO: Min cell voltage: %u:%f", self.cell_min_no, self.cell_min_voltage)
+        
+        self.max_battery_voltage = 3.65 * self.cell_count
```
Comment (Louisvdw) [inline etc/dbus-serialbattery/sinowealth.py:204]: You already set this in get_settings() function and it does not change. Not sure why it is here in cell data as well?

File: etc/dbus-serialbattery/sinowealth.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+          self.read_pack_config_data()
+          
+        cell_index = 1
+        while cell_index <= self.cell_count:
+          self.cell_voltages[cell_index] = self.read_cell_voltage(cell_index)
+          if self.cell_voltages[cell_index] is False:
+            return False
+          cell_index += 1
+        
+        self.cell_max_no = max(self.cell_voltages, key=int) - 1
```
Comment (Louisvdw) [inline etc/dbus-serialbattery/sinowealth.py:196]: If you use the cells object that exist in battery.py instead of your own cell_voltages then the min/max values will be done for you in battery as well.

File: etc/dbus-serialbattery/sinowealth.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        temp_sens_mask = int(~(1 << 6))
+        self.temp_sensors = 1 if (pack_config_data[1] & temp_sens_mask) else 2 # one means two
+        logger.info(">>> INFO: Number of temperatur sensors: %u", self.temp_sensors)
+        return True
+        
+    def read_cell_data(self):
+        if self.cell_count is None:
+          self.read_pack_config_data()
+          
+        cell_index = 1
```
Comment (Louisvdw) [inline etc/dbus-serialbattery/sinowealth.py:189]: Rather use the existing cell object from battery.py for the cell data

**Review by Louisvdw — CHANGES_REQUESTED** [review]

**Comment by Carstijn** [comment]
@Louisvdw thanks for the review, valid point! I will make the changes and try to test them before posting again.

**Comment by Louisvdw** [comment]
@Carstijn 
Here is an example of how your read_cell_data() function can be changed to use the Cell object (for some reason I get a github error page trying to create a pull request on your file/repo.

```
def read_cell_data(self):
	if self.cell_count is None:
	  self.read_pack_config_data()
	# clear the list
	for c in self.cells:
		self.cells.remove(c)
	# get up to the first 16 cells
	for c in range(self.cell_count):
		cell = Cell(False)
		cell.voltage = self.read_cell_voltage(c+1)
		if self.cell_voltages[cell_index] is False:
			return False
		self.cells.append(cell)
	return True
```

**Comment by Carstijn** [comment]
@Louisvdw I made some changes, to fix your comments, pleas re-review.

About the list, creating and deleting it does not feel very optimal, do you have something about my latest solution? (creating once in settings and only updating)

**Review by Louisvdw — APPROVED** [review]
Great work adding the Sinowealth Daly
